### PR TITLE
Feature/save to specific album

### PR DESCRIPTION
**Pull request by SimonErm**
The rebased version of #53. I messed up rebase last time so i created a new remote branch and this PR.

### CODE REVIEW

**Comment by SimonErm** [comment]
#56

**Comment by SimonErm** [comment]
Is this going to be merged?

**Comment by campezzi** [comment]
Also interested in seeing this merged; looking back at #53, I see there were some changes requested (minor formatting stuff, plus a bigger one regarding providing `options` as an optional third parameter to `saveToCameraRoll` instead of replacing the second one, which would be a breaking change. So I guess that's what's holding it back.

I'd be happy to help get those last changes across the line so it can be reviewed again; let me know if you want/need help.

Cheers!

**Comment by campezzi** [comment]
Also looking forward to this being merged.

**Comment by SimonErm** [comment]
All requests from from #53 are done like format and avoid the breaking changes.

**Comment by reyalpsirc** [comment]
@SimonErm Can't this be changed so that the 2nd parameter can either be a string or an object or maybe have a second method called *saveToCameraRollWithOptions* so that this does not become a breaking change?

File: android/src/main/java/com/reactnativecommunity/cameraroll/CameraRollModule.java [hunk 1]
```diff
@@ -110,21 +110,40 @@ public void saveToCameraRoll(String uri, String type, Promise promise) {
     private final Context mContext;
     private final Uri mUri;
     private final Promise mPromise;
+    private final ReadableMap mOptions;
 
-    public SaveToCameraRoll(ReactContext context, Uri uri, Promise promise) {
+    public SaveToCameraRoll(ReactContext context, Uri uri,ReadableMap options, Promise promise) {
```
Comment (bartolkaruza) [inline android/src/main/java/com/reactnativecommunity/cameraroll/CameraRollModule.java]: formatting

File: android/src/main/java/com/reactnativecommunity/cameraroll/CameraRollModule.java [hunk 2]
```diff
@@ -100,8 +100,8 @@ public String getName() {
    * @param promise to be resolved or rejected
    */
   @ReactMethod
-  public void saveToCameraRoll(String uri, String type, Promise promise) {
-    new SaveToCameraRoll(getReactApplicationContext(), Uri.parse(uri), promise)
+  public void saveToCameraRoll(String uri, ReadableMap options, Promise promise) {
+    new SaveToCameraRoll(getReactApplicationContext(), Uri.parse(uri),options, promise)
```
Comment (bartolkaruza) [inline android/src/main/java/com/reactnativecommunity/cameraroll/CameraRollModule.java]: formatting

File: android/src/main/java/com/reactnativecommunity/cameraroll/CameraRollModule.java [hunk 3]
```diff
@@ -110,21 +110,40 @@ public void saveToCameraRoll(String uri, String type, Promise promise) {
...
 
     @Override
     protected void doInBackgroundGuarded(Void... params) {
       File source = new File(mUri.getPath());
       FileChannel input = null, output = null;
       try {
-        File exportDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM);
-        exportDir.mkdirs();
+        File environment;
+        if("mov".equals(mOptions.getString("type"))){
```
Comment (bartolkaruza) [inline android/src/main/java/com/reactnativecommunity/cameraroll/CameraRollModule.java]: formatting

File: android/src/main/java/com/reactnativecommunity/cameraroll/CameraRollModule.java [hunk 4]
```diff
@@ -110,21 +110,40 @@ public void saveToCameraRoll(String uri, String type, Promise promise) {
...
+          environment = Environment.getExternalStoragePublicDirectory(
+                  Environment.DIRECTORY_MOVIES);
+        }else{
+          environment = Environment.getExternalStoragePublicDirectory(
+                  Environment.DIRECTORY_PICTURES);
+        }
+        File exportDir;
+        if(!"".equals(mOptions.getString("album"))){
```
Comment (bartolkaruza) [inline android/src/main/java/com/reactnativecommunity/cameraroll/CameraRollModule.java]: formatting

File: ios/RNCCameraRollManager.m [hunk 5]
```diff
@@ -26,16 +26,16 @@
 @implementation RCTConvert (PHAssetCollectionSubtype)
 
 RCT_ENUM_CONVERTER(PHAssetCollectionSubtype, (@{
-   @"album": @(PHAssetCollectionSubtypeAny),
```
Comment (bartolkaruza) [inline ios/RNCCameraRollManager.m]: formatting

File: ios/RNCCameraRollManager.m [hunk 6]
```diff
@@ -46,7 +46,7 @@ + (PHFetchOptions *)PHFetchOptionsFromMediaType:(NSString *)mediaType
 {
   // This is not exhaustive in terms of supported media type predicates; more can be added in the future
   NSString *const lowercase = [mediaType lowercaseString];
-
+
```
Comment (bartolkaruza) [inline ios/RNCCameraRollManager.m]: formatting

File: ios/RNCCameraRollManager.m [hunk 7]
```diff
@@ -98,62 +98,96 @@ static void requestPhotoLibraryAccess(RCTPromiseRejectBlock reject, PhotosAuthor
...
-      saveBlock();
+      saveWithOptions();
     } else {
       [self.bridge.imageLoader loadImageWithURLRequest:request callback:^(NSError *error, UIImage *image) {
         if (error) {
           reject(kErrorUnableToLoad, nil, error);
           return;
         }
-
+
```
Comment (bartolkaruza) [inline ios/RNCCameraRollManager.m]: don't add spaces on empty lines

**Comment by MaximusBaton** [comment]
how to thumb up for this request to be merged? :)
really need it

File: ios/RNCCameraRollManager.m (hunk 7 above)
Comment (SimonErm) [reply ios/RNCCameraRollManager.m]: Maybe we should provide formatter configs with EditorConfig or specific configs for AndroidStudio and XCode like it's done with prettier. I used the XCode formatter previosly, which also inserts the whitespaces. When i use AndroidStudio default, git blame would become useless.

**Comment by MaximusBaton** [comment]
I have just tested this pr, which I manually merged with latest 'mater' files and it worked really fine on both iOS and android.

**Comment by perryfaro** [comment]
@bartolkaruza any update on this pull request?

**Comment by jamajuice** [comment]
Tested and working on iOS 12, Thanks! Merge please

File: ios/RNCCameraRollManager.m [hunk 8]
```diff
@@ -333,19 +366,19 @@ static void RCTResolvePromise(RCTPromiseResolveBlock resolve,
 {
   NSArray<NSURL *> *assets_ = [RCTConvert NSURLArray:assets];
   [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^{
-      PHFetchResult<PHAsset *> *fetched =
```
Comment (bartolkaruza) [inline ios/RNCCameraRollManager.m]: I think this change is only formatting change, please remove all changes that are non-functional

File: ios/RNCCameraRollManager.m [hunk 9]
```diff
@@ -98,49 +97,83 @@ static void requestPhotoLibraryAccess(RCTPromiseRejectBlock reject, PhotosAuthor
...
+  void (^saveWithOptions)(void) = ^void() {
+    if (![options[@"album"] isEqualToString:@""]) {
+  
+      PHFetchOptions *fetchOptions = [[PHFetchOptions alloc] init];
+      fetchOptions.predicate = [NSPredicate predicateWithFormat:@"title = %@", options[@"album"] ];
+      collection = [PHAssetCollection fetchAssetCollectionsWithType:PHAssetCollectionTypeAlbum
+                                                            subtype:PHAssetCollectionSubtypeAny
+                                                            options:fetchOptions].firstObject;
+      // Create the album
+      if (!collection){
```
Comment (bartolkaruza) [inline ios/RNCCameraRollManager.m]: space after )

File: ios/RNCCameraRollManager.m [hunk 10]
```diff
@@ -98,49 +97,83 @@ static void requestPhotoLibraryAccess(RCTPromiseRejectBlock reject, PhotosAuthor
...
+          if (success){
+            PHFetchResult *collectionFetchResult = [PHAssetCollection fetchAssetCollectionsWithLocalIdentifiers:@[placeholder.localIdentifier]
+                                                                                                        options:nil];
+            collection = collectionFetchResult.firstObject;
+            saveBlock();
+          }else{
+            reject(kErrorUnableToSave, nil, error);
+          }
+        }];
+      }else{
```
Comment (bartolkaruza) [inline ios/RNCCameraRollManager.m]: space before else and after else

File: ios/RNCCameraRollManager.m [hunk 11]
```diff
@@ -98,49 +97,83 @@ static void requestPhotoLibraryAccess(RCTPromiseRejectBlock reject, PhotosAuthor
...
+        saveBlock();
+      }
+    }else{
```
Comment (bartolkaruza) [inline ios/RNCCameraRollManager.m]: idem

File: ios/RNCCameraRollManager.m [hunk 12]
```diff
@@ -98,49 +97,83 @@ static void requestPhotoLibraryAccess(RCTPromiseRejectBlock reject, PhotosAuthor
...
       } else {
-        changeRequest = [PHAssetChangeRequest creationRequestForAssetFromImage:inputImage];
+        assetRequest = [PHAssetChangeRequest creationRequestForAssetFromImage:inputImage];
       }
-
-      placeholder = [changeRequest placeholderForCreatedAsset];
-    } completionHandler:^(BOOL success, NSError * _Nullable error) {
-      if (success) {
+      placeholder = [assetRequest placeholderForCreatedAsset];
+      if(![options[@"album"] isEqualToString:@""]){
```
Comment (bartolkaruza) [inline ios/RNCCameraRollManager.m]: space after )

File: ios/RNCCameraRollManager.m [hunk 13]
```diff
@@ -98,49 +97,83 @@ static void requestPhotoLibraryAccess(RCTPromiseRejectBlock reject, PhotosAuthor
...
+        photosAsset = [PHAsset fetchAssetsInAssetCollection:collection options:nil];
+        PHAssetCollectionChangeRequest *albumChangeRequest = [PHAssetCollectionChangeRequest changeRequestForAssetCollection:collection assets:photosAsset];
+        [albumChangeRequest addAssets:@[placeholder]];
+      }
+    } completionHandler:^(BOOL success, NSError *error) {
+      if (success){
         NSString *uri = [NSString stringWithFormat:@"ph://%@", [placeholder localIdentifier]];
         resolve(uri);
-      } else {
+      }else{
```
Comment (bartolkaruza) [inline ios/RNCCameraRollManager.m]: space after } and before {

File: ios/RNCCameraRollManager.m [hunk 14]
```diff
@@ -98,49 +97,83 @@ static void requestPhotoLibraryAccess(RCTPromiseRejectBlock reject, PhotosAuthor
...
-    } completionHandler:^(BOOL success, NSError * _Nullable error) {
-      if (success) {
+      placeholder = [assetRequest placeholderForCreatedAsset];
+      if(![options[@"album"] isEqualToString:@""]){
+        photosAsset = [PHAsset fetchAssetsInAssetCollection:collection options:nil];
+        PHAssetCollectionChangeRequest *albumChangeRequest = [PHAssetCollectionChangeRequest changeRequestForAssetCollection:collection assets:photosAsset];
+        [albumChangeRequest addAssets:@[placeholder]];
+      }
+    } completionHandler:^(BOOL success, NSError *error) {
+      if (success){
```
Comment (bartolkaruza) [inline ios/RNCCameraRollManager.m]: space after )

**Comment by bartolkaruza** [comment]
ok, did another review. There are still issues with the formatting, otherwise LGTM :)

File: ios/RNCCameraRollManager.m (hunk 8 above)
Comment (SimonErm) [reply ios/RNCCameraRollManager.m]: Are you joking? :D The last two Reviews, you ask me to fix formatting.

**Comment by MaximusBaton** [comment]
Hey, @SimonErm today I was checking this feature again in my project (manually merged the changes) and noticed that photos are being added in both **Photos->Albums->MyAlbums->_custom_** album and iPhone's **Photos->Albums->_Camera Roll_** album. So, duplication occurs.

Is that a normal behavior or smth has changed? [Don't remember this to be happening when checking the pull request. correct me if I'm wrong]

Latest macOS, latest XCode, iPhone simulator 5s

**Comment by SimonErm** [comment]
@MaximusBaton To be honest i can't remember :-D I will take a look at it the next days.

**Comment by MaximusBaton** [comment]
@SimonErm , I removed photos from iPhone's **Photos->Albums->_Camera Roll_** album and they disappeared from custom album. Which means, no duplication exist!!! :)

I'm not a professional iPhone user, but looks like all fine. iPhone just collects all the photos at one 'Camera roll' album and then on the same page shows them in separate custom albums. weird :)

cheers ))

File: ios/RNCCameraRollManager.m (hunk 8 above)
Comment (bartolkaruza) [reply ios/RNCCameraRollManager.m]: I meant fix the formatting to match the *current* formatting. This is a change to your formatting style

**Comment by bartolkaruza** [comment]
Ok, I tried to clean up the last wonky formatting issues myself. How do iOS devs *survive* without prettier? @SimonErm Thank you for the work!

**Comment by SimonErm** [comment]
@bartolkaruza thank you 😊 i don't know 😆 maybe we should use clang formatter or change formatting to the xcode default.

**Pull request merged**